### PR TITLE
SUBMARINE-679. Spark Security Plugin Support Spark-3.0.1

### PR DESCRIPTION
**Pull request by atovk**
### What is this PR for?
Submarine Spark Security Plugin Support Spark-3.x

### What type of PR is it?
[Improvement | Feature]

### What is the Jira issue?
[SUBMARINE-679](https://issues.apache.org/jira/browse/SUBMARINE-679)

### Questions:
* Does this need documentation? YES

### CODE REVIEW

**Comment by atovk** [comment]
@yaooqinn This test was Cancelled in cicd, Need to be processed?

**Comment by yaooqinn** [comment]
> @yaooqinn This test was Cancelled in cicd, Need to be processed?

that's fine.

**Comment by yaooqinn** [comment]
we need to add CI for spark-3 profile

**Comment by atovk** [comment]
> we need to add CI for spark-3 profile

ok

**Comment by yaooqinn** [comment]
https://github.com/apache/submarine/blob/master/.travis.yml#L257-L335 FYI

**Comment by xunliu** [comment]
@yaooqinn Please help review this PR again. :-)

**Review by xunliu — COMMENTED** [review]
@yaooqinn Please help review this PR again. :-)

**Comment by atovk** [comment]
@yaooqinn Spark Security Plugin Support Spark-3.0.1 unit test passed, but has other module case failure

File: submarine-security/spark-security/pom.xml [hunk 1]
```diff
@@ -539,6 +563,7 @@
       <id>spark-2.3</id>
       <properties>
         <spark.version>2.3.4</spark.version>
+        <guava.version>15.0</guava.version>
```
Comment (yaooqinn) [inline submarine-security/spark-security/pom.xml:566]: can you explain why we need to add this?

File: submarine-security/spark-security/pom.xml [hunk 2]
```diff
@@ -547,10 +572,25 @@
...
         <scalatest.version>3.0.3</scalatest.version>
       </properties>
     </profile>
 
+    <profile>
+      <id>spark-3.0</id>
+      <properties>
+        <spark.version>3.0.1</spark.version>
+        <scala.version>2.12.12</scala.version>
```
Comment (yaooqinn) [inline submarine-security/spark-security/pom.xml:584]: Spark is compiled w/ scala 2.12.10, can we make it consistent w/ spark?

File: submarine-security/spark-security/spark-2/src/main/scala/org/apache/submarine/spark/security/parser/SubmarineSqlParserCompatible.scala [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.submarine.spark.security.parser
+
+import org.apache.spark.sql.catalyst.parser.ParserInterface
+
+class SubmarineSqlParserCompatible(override val delegate: ParserInterface) extends SubmarineSqlParser(delegate: ParserInterface) {
```
Comment (yaooqinn) [inline submarine-security/spark-security/spark-2/src/main/scala/org/apache/submarine/spark/security/parser/SubmarineSqlParserCompatible.scala:24]: what is this for?

File: submarine-security/spark-security/pom.xml (hunk 2 above)
Comment (atovk) [reply submarine-security/spark-security/pom.xml:584]: sure, i well be change to `scala 2.12.10`

File: submarine-security/spark-security/pom.xml (hunk 1 above)
Comment (atovk) [reply submarine-security/spark-security/pom.xml:566]: this change for test case in profile  hadoop-3.x spark-3.x, case runner has guava error 
`java.lang.NoSuchMethodError: com.google.common.base.Preconditions.checkArgument(ZLjava/lang/String;Ljava/lang/Object;)V`
Dependency guava version is old; 
so, i add guava Dependency for test scope, profile `hadoop-3.2, ranger-x, spark-3.0 ` test case passed, 
but profile `hadoop-2.9, ranger-x, spark-2.x` test case is failure [Cause: java.lang.IllegalAccessError: tried to access method com.google.common.base.Stopwatch.<init>](https://stackoverflow.com/questions/36427291/illegalaccesserror-to-guavas-stopwatch-from-org-apache-hadoop-mapreduce-lib-inp) , spark-2.x need guava version is 15.0;

if skip profile `hadoop-3.x` test case, i well be clean it

File: submarine-security/spark-security/spark-2/src/main/scala/org/apache/submarine/spark/security/parser/SubmarineSqlParserCompatible.scala (hunk 3 above)
Comment (atovk) [reply submarine-security/spark-security/spark-2/src/main/scala/org/apache/submarine/spark/security/parser/SubmarineSqlParserCompatible.scala:24]: in spark-3.x `ParserInterface` interface added 2 method : `parseMultipartIdentifier` and  `parseRawDataType`, this change for compatible spark2.x and spark3.x

Comment (yaooqinn) [reply submarine-security/spark-security/spark-2/src/main/scala/org/apache/submarine/spark/security/parser/SubmarineSqlParserCompatible.scala:24]: `SubmarineSqlParser` is a public API, it's not a good approach here

Comment (atovk) [reply submarine-security/spark-security/spark-2/src/main/scala/org/apache/submarine/spark/security/parser/SubmarineSqlParserCompatible.scala:24]: if  `SubmarineSqlParser ` is public , has someone used it, we can Compatible `SubmarineSqlParser` class, not subclass extends it ?
move class ` SubmarineSqlParser` to spark2.x and spark3.x, has different implementations : 
spark3.x add method : `parseMultipartIdentifier` and `parseRawDataType`;
spark2.x same with `org.apache.submarine.spark.security.parser.SubmarineSqlParser`

this approach better, any have other idea?

**Comment by atovk** [comment]
@yaooqinn review it thx!

File: submarine-security/spark-security/spark-3/src/main/scala/org/apache/submarine/spark/compatible/command/CompatibleCommand.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+package org.apache.submarine.spark.compatible
+
+import org.apache.spark.sql.catalyst.analysis.PersistedView
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, SetCatalogAndNamespace, ShowNamespaces, Subquery}
+
+
+package object CompatibleCommand {
+
+  type ShowDatabasesCommandCompatible = ShowNamespaces
```
Comment (yaooqinn) [inline submarine-security/spark-security/spark-3/src/main/scala/org/apache/submarine/spark/compatible/command/CompatibleCommand.scala:28]: Can we make sure that the output is not changed?

**Comment by yaooqinn** [comment]
looks ok to me in general.

File: submarine-security/spark-security/spark-3/src/main/scala/org/apache/submarine/spark/compatible/command/CompatibleCommand.scala (hunk 4 above)
Comment (atovk) [reply submarine-security/spark-security/spark-3/src/main/scala/org/apache/submarine/spark/compatible/command/CompatibleCommand.scala:28]: yes, this change is work well with hive metadata, output not change;

**Comment by yaooqinn** [comment]
cc @xunliu, we may need a long way to make full support for spark 3.0 but this could be a good start. thanks @atovk

**Review by xunliu — APPROVED** [review]
@atovk Thank you contribution this feature.
Will merge if no more comments.

**Comment by atovk** [comment]
@yaooqinn @xunliu thx!

**Pull request closed without merging**